### DISCUSSION TITLE
Is the MobileFormatter able to support object graphs with circular references?

### DISCUSSION THREAD

**Discussion by j055** [MarimerLLC/csla#2387] [Questions] [answered]
Instances of the class below throw a `StackOverflow` exception:

```
[Serializable]
public sealed class Circle : MobileObject
{
    public MobileList<Circle> Children { get; set; }

    protected override void OnGetState(SerializationInfo info, StateMode mode)
    {
        base.OnGetState(info, mode);
        info.AddValue(nameof(Children), Children);
    }

    protected override void OnSetState(SerializationInfo info, StateMode mode)
    {
        base.OnSetState(info, mode);
        Children = info.GetValue<MobileList<Circle>>(nameof(Children));
    }
}
```

```
[TestMethod]
public void CircularMobileFormatterTest()
{
    var obj = new Circle();
    obj.Children = new MobileList<Circle> { obj };

    var array = MobileFormatter.Serialize(obj);  // throws `StackOverflow` eventually
    var result = (Circle)MobileFormatter.Deserialize(array);

    Assert.AreEqual(1, result.Children.Count);
}
```

The `BinaryFormatter` works fine so I'm guessing I need to work on the implementation of the `OnGetChildren` method but I've no idea what I need to do or even if it can work.

Can you point me in the right direction?
Thanks
Andrew

**Comment by rockfordlhotka** [comment, marked as answer]
Right. The `OnGetState` and `OnSetState` methods are for _instance values only_, not references to other objects in the graph. For references you _must_ implement the get/set children methods.

Look at [Csla.Core.MobileList](https://github.com/MarimerLLC/csla/blob/main/Source/Csla/Core/MobileList.cs) - a type that only has children state, and no state of its own.
